x-pack/plugins/enterprise_search/server/collectors/app_search/telemetry.ts

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { set } from 'lodash';
+import {
+  ISavedObjectsRepository,
+  SavedObjectsServiceStart,
+  SavedObjectAttributes,
+} from 'src/core/server';
+import { UsageCollectionSetup } from 'src/plugins/usage_collection/server';
+
+interface ITelemetry {
+  ui_viewed: {
+    setup_guide: number;
+    engines_overview: number;
+  };
+  ui_error: {
+    cannot_connect: number;
+  };
+  ui_clicked: {
+    create_first_engine_button: number;
+    header_launch_button: number;
+    engine_table_link: number;
+  };
+}
+
+export const AS_TELEMETRY_NAME = 'app_search_telemetry';
+
+/**
+ * Register the telemetry collector
+ */
+
+export const registerTelemetryUsageCollector = (
+  usageCollection: UsageCollectionSetup,
+  savedObjects: SavedObjectsServiceStart
+) => {
+  const telemetryUsageCollector = usageCollection.makeUsageCollector<ITelemetry>({
+    type: 'app_search',
+    fetch: async () => fetchTelemetryMetrics(savedObjects),
+    isReady: () => true,
+    schema: {
+      ui_viewed: {
+        setup_guide: { type: 'long' },
+        engines_overview: { type: 'long' },
+      },
+      ui_error: {
+        cannot_connect: { type: 'long' },
+      },
+      ui_clicked: {
+        create_first_engine_button: { type: 'long' },
+        header_launch_button: { type: 'long' },
+        engine_table_link: { type: 'long' },
+      },
+    },
+  });
+  usageCollection.registerCollector(telemetryUsageCollector);
+};
+
+/**
+ * Fetch the aggregated telemetry metrics from our saved objects
+ */
+
+const fetchTelemetryMetrics = async (savedObjects: SavedObjectsServiceStart) => {
+  const savedObjectsRepository = savedObjects.createInternalRepository();
+  const savedObjectAttributes = (await getSavedObjectAttributesFromRepo(
+    savedObjectsRepository
+  )) as SavedObjectAttributes;
+
+  const defaultTelemetrySavedObject: ITelemetry = {
+    ui_viewed: {
+      setup_guide: 0,
+      engines_overview: 0,
+    },
+    ui_error: {
+      cannot_connect: 0,
+    },
+    ui_clicked: {
+      create_first_engine_button: 0,
+      header_launch_button: 0,
+      engine_table_link: 0,
+    },
+  };
+
+  // If we don't have an existing/saved telemetry object, return the default
+  if (!savedObjectAttributes) {
+    return defaultTelemetrySavedObject;
+  }
+
+  // Iterate through each attribute key and set its saved values
+  const attributeKeys = Object.keys(savedObjectAttributes);
+  const telemetryObj = defaultTelemetrySavedObject;
+  attributeKeys.forEach((key: string) => {
+    set(telemetryObj, key, savedObjectAttributes[key]);
```

Comment:
Ahhh yeah! The upgrade assistant is definitely where I grabbed this code from. This makes a ton of sense - I think originally tried to get rid of the `set` dependency as well but because of the way the key strings were being read (something like `'parentkey.childkey'` IIRC I got kinda lazy and left in. Looking at the changes now!

Reply:
Addressed in https://github.com/elastic/kibana/pull/66922/commits/f4ae1f221b629b22a4bd2107ba0cefd2d1b96382